enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,358 @@
+
+On the OVN side:
+- OVN accepts the packet via the VXLAN tunnel, ignore the VNID set by SDN and
+  then just routes it normally.
+- Ingress NetworkPolicy processing will happen when the packet reaches the
+  destination pod’s switch port, just like normal.
+  - Our NetworkPolicy rules are all based on IP addresses, not “logical input
+    port”, etc, so it doesn’t matter that the packets came from outside OVN and
+    have no useful OVN metadata.
+
+#### Packets going from OVN Kubernetes to OpenShift SDN
+
+On the ovn side:
+- The packet just follow the same path of hybrid overlay. The packet just has to
+  get routed out the VXLAN tunnel with VNID 0.
+
+On the sdn side:
+- We have to change ingress NetworkPolicy processing to be based entirely on pod
+  IPs rather than using namespace VNIDs, since packets from ovn nodes won’t have
+  the VNID set. There is already code to generate the rules that way though,
+  because egress NetworkPolicy already works that way.
+
+### The Migration Process
+
+#### Migration Setup
+
+1. The admin kicks off the migration process by editing the CNO configuration
+
+2. CNO will label each existing node with, eg:
+
+    ```yaml
+    migration.network.openshift.io/plugin: OpenShiftSDN
+    k8s.ovn.org/hybrid-overlay-node-subnet: 10.129.0.0/23
+    ```
+
+   (CNO knows how to figure out the host subnet by looking at openshift-sdn
+   HostSubnet objects. When migrating in the reverse direction, it knows to look
+   at the ovn-kubernetes annotations.)
+
+3. CNO will redeploy the openshift-sdn DaemonSets, enabling “migration mode”,
+   and adding node affinity to the node DaemonSet:
+    
+    ```yaml
+    affinity:
+      nodeAffinity:
+        requiredDuringSchedulingIgnoredDuringExecution:
+          nodeSelectorTerms:
+            - matchExpressions:
+                - key: migration.network.openshift.io/plugin
+                  operator: In
+                  values: ["OpenShiftSDN"]
+    ```
+
+  so that openshift-sdn will now only run on nodes labeled
+  “migration.network.openshift.io/plugin: OpenShiftSDN” (which is currently all
+  of them).
+
+4. CNO will then deploy the ovnkube-master and ovnkube-node DaemonSets, also
+   enabling hybrid overlay mode” in their config, and adding equivalent node
+   affinity to the ovnkube-node DaemonSet (to only run on nodes labeled
+   “migration.network.openshift.io/plugin: OVNKubernetes”). Since there are not
+   currently any nodes labeled that way, this means no actual ovnkube-node pods
+   will be created at this step.
+
+At this point, neither plugin is willing to run on a node that has no
+“migration.network.openshift.io/plugin” label at all. This means that if a new
+node gets added to the cluster in the middle of the migration process, neither
+plugin will get deployed to it until after the migration of the older nodes is
+complete, and the node affinity is removed.
+
+#### Migration
+
+In order to run ovn-kubernetess in shared-gw mode, we need MCO to update the
```

Comment:
OK a lot going on here.  Would it help if we had the SDN daemonset actually *live reconfigure* the service?  That's fully supported by systemd and the MCO.